gh-89730: EncryptedClientHello support in ssl module

irl opened the pull request:
Exposes options for clients to use Encrypted Client Hello (ECH) when establishing TLS connections. It is up to the user to source the ECH public keys before making use of these options.

To use these options requires a version of openssl that includes these options, which currently are on a feature branch:

https://github.com/openssl/openssl/tree/feature/ech

We do not expect that the API here will have any substaintial changes when it is merged.

For testing, clone the above repository, checkout the `feature/ech` branch, and build. Then use this openssl installation when building cPython:

```
git clone https://github.com/openssl/openssl.git
pushd openssl
git checkout feature/ech
./configure --prefix=$HOME/opt/openssl
make -j8
make install_sw
popd
git clone https://github.com/irl/cpython.git
git checkout issue-89730
./configure --with-pydebug --with-openssl=$HOME/opt/openssl
make -j8
```

Additionally I have added to the documentation an example of the use of these options, which is built at: https://irl.github.io/cpython/library/ssl.html#encrypted-client-hello

<!-- gh-issue-number: gh-89730 -->
* Issue: gh-89730
<!-- /gh-issue-number -->


<!-- readthedocs-preview cpython-previews start -->
----
📚 Documentation preview 📚: https://cpython-previews--135435.org.readthedocs.build/

<!-- readthedocs-preview cpython-previews end -->

ZeroIntensity commented on Modules/_ssl.c:
```diff
@@ -2073,6 +2077,111 @@ cipher_to_dict(const SSL_CIPHER *cipher)
+
+    int status = SSL_ech_get1_status(self->ssl, &inner_sni, &outer_sni);
+	PyObject *retval = PyTuple_New(3);
+    PyTuple_SetItem(retval, 0, PyLong_FromLong(status));
```
`PyTuple_New` can fail, and use `PyTuple_SET_ITEM` to avoid error checking there.

irl replied:
Now testing return value for PyTuple_New and using PyTuple_SET_ITEM here.

ZeroIntensity commented on Modules/_ssl.c:
```diff
@@ -2073,6 +2077,111 @@ cipher_to_dict(const SSL_CIPHER *cipher)
+    PyTuple_SetItem(retval, 0, PyLong_FromLong(status));
+
+    if (inner_sni != NULL) {
+        PyTuple_SetItem(retval, 1, PyUnicode_FromString(inner_sni));
```
`PyUnicode_FromString` can fail.

irl replied:
Checking return value, cleaning up if failed.

ZeroIntensity commented on Modules/_ssl.c:
```diff
@@ -2073,6 +2077,111 @@ cipher_to_dict(const SSL_CIPHER *cipher)
        );
 }
 
+/*[clinic input]
```
I'm pretty sure you need to decorate these with `@critical_section` for free-threading.

irl replied:
Added.

ZeroIntensity commented on Modules/_ssl.c:
```diff
@@ -70,6 +70,10 @@
 #include "openssl/bio.h"
 #include "openssl/dh.h"
 
+#if __has_include("openssl/ech.h")
```
Check `#ifdef __has_include`.

picnixz replied:
Also use the macro `OPENSSL_ECH` that you defined (which should be renamed)

picnixz commented on Doc/library/ssl.rst:
```diff
@@ -214,7 +214,6 @@ purposes.
       The context now uses :data:`VERIFY_X509_PARTIAL_CHAIN` and
       :data:`VERIFY_X509_STRICT` in its default verify flags.
 
-
```
Keep this line.

picnixz commented on Doc/library/ssl.rst:
```diff
@@ -1003,6 +1002,13 @@ Constants
+   :class:`enum.IntEnum` collection of Encrypted Client Hello (ECH) statuses
+   returned by :meth:`SSLSocket.get_ech_status`.
+
+   .. versionadded:: TODO XXX
```
```suggestion
   .. versionadded:: next
```

And ditto for the rest

picnixz commented on Misc/NEWS.d/next/Library/2025-06-12-15-39-48.gh-issue-89730.Lu35Fu.rst:
```diff
@@ -0,0 +1,8 @@
+no public key is available, an option is available to "GREASE" the
+connection, and it is possible to retrieve the public key from the retry
+configuration sent by servers that support ECH as they terminate the initial
+connection.
```
This description should also be put in What's New rather here.

irl replied:
I don't know what this means.

picnixz commented on Misc/NEWS.d/next/Library/2025-06-12-15-39-48.gh-issue-89730.Lu35Fu.rst:
```diff
@@ -0,0 +1,8 @@
+Adds support for Encrypted Client Hello (ECH) to the ssl module. Clients may
+use the options exposed to establish TLS connections using ECH. Third-party
+libraries like dnspython can be used to query for HTTPS and SVCB records
```
```suggestion
libraries like ``dnspython`` can be used to query for HTTPS and SVCB records
```

picnixz commented on configure.ac:
```diff
@@ -7551,6 +7551,29 @@ WITH_SAVE_ENV([
+if test "$have_openssl_ech_h" = "yes"; then
+    AC_MSG_CHECKING([for OpenSSL ECH support])
+    AC_CHECK_DECLS([SSL_ECH_STATUS_GREASE_ECH],
+                   [AC_DEFINE([OPENSSL_ECH], [1],
```
Can we use `HAVE_OPENSSL_ECH` naming as for other macros please? See the AC func PY_CHECK_FUNC.

irl replied:
I've updated this to use PY_CHECK_FUNC but honestly I have no idea what I'm doing with autoconf and I could not find any documentation on correct usage of PY_CHECK_FUNC so it may just be working by chance.

picnixz commented on configure.ac:
```diff
@@ -7551,6 +7551,29 @@ WITH_SAVE_ENV([
+                              [Define if OpenSSL does not have ECH support])],
+                   [
+		     #include <openssl/ech.h>
+		     ])
```
Avoid using tabs.

picnixz commented on Modules/_ssl.c:
```diff
@@ -2073,6 +2077,111 @@ cipher_to_dict(const SSL_CIPHER *cipher)
        );
 }
 
+/*[clinic input]
```
This needs to be guarded by an ifdef because OpenSSL may be built without ECH support: https://github.com/openssl/openssl/compare/master...feature/ech#diff-6defafa2caa65304173956e3fbc26b6947c50bc769c08c3da32c1032fa6519d4R473.

picnixz commented on Modules/_ssl.c:
```diff
@@ -2073,6 +2077,111 @@ cipher_to_dict(const SSL_CIPHER *cipher)
+
+    return retval;
+#else
+    PyErr_SetString(PyExc_NotImplementedError, "OpenSSL does not have ECH support");
```
I don't remember whether we raise a NotImplementedError or just plainly don't define the function at the module level.

irl replied:
I see other functions just being undefined e.g.

```
#ifndef OPENSSL_NO_PSK
static unsigned int psk_server_callback(SSL *s,
```

This seems a nice clean way to do it, but right now there wouldn't be an OPENSSL_NO_ECH unless it's a version of OpenSSL that knows about ECH, so should it be two #if for if it knows about ECH and then if ECH is enabled?

picnixz commented on Modules/_ssl.c:
```diff
@@ -2073,6 +2077,111 @@ cipher_to_dict(const SSL_CIPHER *cipher)
+
+    SSL_ech_get1_retry_config(self->ssl, &ec, &eclen);
+
+    return PyBytes_FromStringAndSize((char *)ec, eclen);
```
```suggestion
    if (self->ssl == NULL) {
        Py_RETURN_NONE;
    }
    SSL_ech_get1_retry_config(self->ssl, &ec, &eclen);
    return PyBytes_FromStringAndSize((char *)ec, eclen);
```

picnixz commented on Modules/_ssl.c:
```diff
@@ -2073,6 +2077,111 @@ cipher_to_dict(const SSL_CIPHER *cipher)
+
+    if (outer_server_name != NULL) {
+        result = SSL_ech_set1_outer_server_name(self->ssl, outer_server_name, 0);
+    } else {
```
```suggestion
    }
    else {
```

Please follow PEP-7

picnixz commented on Modules/_ssl.c:
```diff
@@ -3472,6 +3584,43 @@ _ssl__SSLContext__set_alpn_protocols_impl(PySSLContext *self,
+    }
+    memcpy(self->alpn_protocols, protos->buf, protos->len);
+    self->alpn_protocols_len = (unsigned int)protos->len;
+    if (SSL_CTX_ech_set1_outer_alpn_protos(self->ctx, self->alpn_protocols, self->alpn_protocols_len)) {
```
Wrap lines under 80 chars

picnixz commented on Modules/_ssl.c:
```diff
@@ -4600,6 +4749,45 @@ _ssl__SSLContext_set_default_verify_paths_impl(PySSLContext *self)
+        || (es = OSSL_ECHSTORE_new(NULL, NULL)) == NULL
+        || OSSL_ECHSTORE_read_echconfiglist(es, es_in) != 1
+        || SSL_CTX_set1_echstore(self->ctx, es) != 1) {
+        _setSSLError(get_state_ctx(self), NULL, 0, __FILE__, __LINE__);
```
Break down this big if into multiple ones and use a goto error with an error label doing the cleanup.

picnixz commented on Lib/ssl.py:
```diff
@@ -459,7 +464,7 @@ def wrap_socket(self, sock, server_side=False,
             server_hostname=server_hostname,
             context=self,
-            session=session
+            session=session,
```
Revert this change, it's not needed.

picnixz commented on Lib/ssl.py:
```diff
@@ -831,6 +833,14 @@ def context(self):
+        if self._sslobj:
+            return self._sslobj.get_ech_status()[2]
+        else:
+            raise ValueError("No SSL wrapper around " + str(self))
```
```suggestion
    @property
    def outer_server_hostname(self):
        """The server name used in the outer ClientHello."""
        return self._sslobj.get_ech_status()[2]
```

We don't use type annotations in the standard library, it's left to typeshed. And let's just propagate an AttributeError as for the other properties.

picnixz commented on Lib/ssl.py:
```diff
@@ -1092,6 +1108,14 @@ def context(self, ctx):
+        if self._sslobj:
+            return self._sslobj.get_ech_status()[2]
+        else:
+            raise ValueError("No SSL wrapper around " + str(self))
```
```suggestion
    def outer_server_hostname(self):
        """The server name used in the outer ClientHello."""
        if self._sslobj:
            return self._sslobj.get_ech_status()[2]
```

Other properties actually return None if sslobj is None.

picnixz commented on Lib/ssl.py:
```diff
@@ -1527,3 +1558,13 @@ def get_server_certificate(addr, ssl_version=PROTOCOL_TLS_CLIENT,
+    protos = bytearray()
+    for protocol in alpn_protocols:
+        b = bytes(protocol, 'ascii')
+        if len(b) == 0 or len(b) > 255:
```
```suggestion
        if not b or len(b) > 255:
```

picnixz reviewed (changes requested):
As you mentioned, the ECH feature is not yet available in OpenSSL (https://github.com/openssl/openssl/tree/feature/ech). I would like to defer this until it's released.

In addition:

- please remove blank lines as much as possile. Too many blank lines hinder the reading flow.
- check return codes from OpenSSL (SSL_ech_get1_retry_config may fail for instance, so you should check if it succeeded, and if not, correctly handle the return code)
- we need tests

irl committed:
EncryptedClientHello support in ssl module

Exposes options for clients to use Encrypted Client Hello (ECH) when
establishing TLS connections. It is up to the user to source the
ECH public keys before making use of these options.

AdamKorcz commented on Modules/_ssl.c:
```diff
@@ -4600,6 +4765,58 @@ _ssl__SSLContext_set_default_verify_paths_impl(PySSLContext *self)
+    if (es_in == NULL) {
+        goto fail;
+    }
+    es = OSSL_ECHSTORE_new(NULL, NULL));
```
Double closing parenthesis looks like a typo.

AdamKorcz commented on Modules/_ssl.c:
```diff
@@ -2073,6 +2077,127 @@ cipher_to_dict(const SSL_CIPHER *cipher)
+    if (retval == NULL) {
+        goto fail;
+    }
+    PyTuple_SET_ITEM(retval, 0, PyLong_FromLong(status));
```
nit: `PyLong_FromLong` can return `NULL`. Should execution continue in that case?

picnixz replied:
No indeed, so this should be checked.

AdamKorcz commented on Modules/_ssl.c:
```diff
@@ -2073,6 +2077,127 @@ cipher_to_dict(const SSL_CIPHER *cipher)
+        Py_INCREF(Py_None);
+    }
+    return retval;
+fail:
```
FWICS `retval` needs to be free'd to in `fail`.

AdamKorcz commented on Doc/library/ssl.rst:
```diff
@@ -2813,6 +2887,52 @@ of TLS/SSL. Some new TLS 1.3 features are not yet available.
+            context.set_ech_config(ech_config)
+        with socket.create_connection((hostname, 443)) as sock:
+            with context.wrap_socket(sock, server_hostname=hostname) as ssock:
+                if (ssock.get_ech_status == ECHStatus.ECH_STATUS_GREASE_ECH
```
This will always resolve to `False` because it is checking `ssock.get_ech_status` as a bound method object and not the result of the method. It should be `ssock.get_ech_status()`.

picnixz commented on Modules/_ssl.c:
```diff
@@ -4600,6 +4765,58 @@ _ssl__SSLContext_set_default_verify_paths_impl(PySSLContext *self)
+/*[clinic end generated code: output=cea53188b338cf80 input=3ef0514cae6ab6ee]*/
+{
+#ifdef HAVE_OPENSSL_ECH
+    BIO *es_in = BIO_new_mem_buf(ech_config->buf, ech_config->len);
```
can we avoid cryptic variable names please?

picnixz commented on Modules/_ssl.c:
```diff
@@ -4600,6 +4765,58 @@ _ssl__SSLContext_set_default_verify_paths_impl(PySSLContext *self)
+    OSSL_ECHSTORE_free(es);
+    BIO_free_all(es_in);
+    PyBuffer_Release(ech_config);
+    Py_RETURN_NONE;
+fail:
```
```suggestion
    Py_RETURN_NONE;

fail:
```

picnixz commented on Modules/_ssl.c:
```diff
@@ -4600,6 +4765,58 @@ _ssl__SSLContext_set_default_verify_paths_impl(PySSLContext *self)
+    }
+    OSSL_ECHSTORE_free(es);
+    BIO_free_all(es_in);
+    PyBuffer_Release(ech_config);
```
AFAIR, there is no need for that. The caller handles it implicitly.

picnixz reviewed (commented):
Considering it's still not released in OpenSSL, I prefer that we close this PR and re-open it once it's shipped. There are also conflicts to solve and the implementation is lacking tests and there is no ETA for that feature (we also need to find another way to configure this, because I'm not really happy with the configure check, we may need more checks with OPENSSL_VERSION for instance).
